README.md
# XOD

[![CircleCI](https://circleci.com/gh/xodio/xod/tree/master.svg?style=shield)](https://circleci.com/gh/xodio/xod/tree/master)

XOD is a visual programming language for microcontrollers. This repository contains sources for XOD language core, XOD IDE and XOD standard library.

![Xoding demo](./.github/xoding.gif)

## Installation & Quick start

Download the latest IDE version for desktop or run the browser-based IDE at <https://xod.io>.

Documentation and tutorials are at <https://xod.io/docs/>.

## Building from source

XOD is written in JavaScript and ReasonML. You need Node.js and Yarn to build from source. Make sure they are available on your system.

Clone the repository and set working directory to its root. Then run:

```bash
# Install all JavaScript and ReasonML dependencies
yarn

# Build all packages of XOD
yarn build
```

To start the desktop IDE run:

```bash
yarn start:electron
```

Alternatively, run browser-based IDE:

```bash
yarn dev:browser
# IDE is available at <http://localhost:8080>
```

## Directory structure

The project is managed as a [Lerna](https://github.com/lerna/lerna) monorepo and split up in few directories:

* `packages/` — most of source code is here; navigate to a particular package to see it’s own `README` and get an idea what it is for
* `tools/` — utility scripts to assist build process and routine maintenance tasks
* `workspace/` — XOD standard library, default projects, and end-to-end fixtures

## Repository commands

You can run several commands on source files. They are available as yarn subcommands:

* `yarn build` — build, transpile, pack all
* `yarn build:electron` — build desktop IDE only
* `yarn build:cli` — build CLI tools only
* `yarn dev:browser` — run dev-version of browser IDE on localhost
* `yarn dist:electron` — build OS-specific distributive of desktop IDE
* `yarn test` — run unit tests
* `yarn test-cpp` — run C++ code tests
* `yarn test-func` — run functional tests
* `yarn tabtest` — run standard library tabular tests
* `yarn lint` — run the linter to check code style
* `yarn verify` — build, lint, test; run this prior to a pull request
* `yarn start:electron` — starts desktop IDE
* `yarn start:spectron-repl` — starts functional tests environment
* `yarn storybook` — starts React components viewer for visual inspection
* `yarn clean` — remove build artifacts and installed `node_modules`

Note that dependencies between tasks are not resolved. `test` and `start:*` expect that the project is already built.

### Scoping

Many commands (notably `build`, `dev`, `test`) support package scoping to save development time. To rebuild only `xod-project`:

```bash
yarn build --scope xod-project
```

To rebuild `xod-project` and its dependencies:

```bash
yarn build --scope xod-project --include-filtered-dependencies
```

Those are standard [Lerna flags](https://github.com/lerna/lerna#flags).

### Debugging functional tests

`yarn test-func` runs automated end-to-end functional tests.

You can set `XOD_DEBUG_TESTS` environment variable to keep IDE open on failure: `XOD_DEBUG_TESTS=1 yarn test-func`

Use `yarn start:spectron-repl` to run an interactive session and control the IDE window programmatically.

### Running C++ and tabular tests

You need `gcc` and `avr-gcc` to be installed system-wide to run C++ code tests. They are available as OS packages for most platforms.

## License

Copyright 2017-2019 XOD Inc.

This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License, version 3, as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License along with this program. If not, see <http://www.gnu.org/licenses/>.

As a special exception, the copyright holders give permission to link the code of portions of this program with the OpenSSL library under certain conditions as described in each individual source file and distribute linked combinations including the program with the OpenSSL library. You must comply with the GNU Affero General Public License in all respects for all of the code used other than as permitted herein. If you modify file(s) with this exception, you may extend this exception to your version of the file(s), but you are not obligated to do so. If you do not wish to do so, delete this exception statement from your version. If you delete this exception statement from all source files in the program, then also delete it in the license file.

## Contributing

Feel free to contribute to the project! See the general [Contibutor’s guide](https://xod.io/docs/contributing/) and [GitHub contribution guidelines](./CONTRIBUTING.md).


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.38.0"></a>

## 0.38.0 (2021-03-12)

### Features and enhancements

* [core] Make possible to pass variadic pins to children variadic nodes (#2107)
* [ide] Add table log feature: collect, view and save tabular data in live session mode (#2098, #2108)
* [nodes] Add nodes to handle interrupts in [xod/gpio](https://xod.io/libs/xod/gpio) standard library (#2106)
* [nodes] Add [xod/core/micros](https://xod.io/libs/xod/core/micros) node and some utilities for this type (#2106)
* [nodes] Add `variadic-pass-*` marker nodes (#2107)
* [nodes] Add [xod/debug/table-log](https://xod.io/libs/xod/debug/table-log) node to collect tabular data (#2098, #2108)

### Bug fixes

* [ide] Fix minor mistakes in welcome-to-xod project (#2096)
* [ide] Prevent corrupting projects when workspace contains libaries with invalid names (#2099)
* [ide] Fix scrollbars in C++ editor (#2104)

<a name="0.37.3"></a>

## 0.37.3 (2021-02-17)

### Bug fixes

* [ide] Prevent toggling helpbox by "h" key when interacting with text inputs (#2101)

<a name="0.37.2"></a>

## 0.37.2 (2021-01-25)

### Bug fixes

* [infra] Fix a bug in the release process that resulted in a bad windows build (#2090)

<a name="0.37.1"></a>

## 0.37.1 (2021-01-20)

### Bug fixes

* [ide] Fix a bug in storing and retrieving settings (#2087)

<a name="0.37.0"></a>

## 0.37.0 (2021-01-18)

### Features and enhancements

* [ide] Make deployment pane resizable (#2081)
* [ide] Copy and save logs from deployment pane (#2078, #2079)
* [ide] Recover IDE from runtime exceptions (#2074)
* [ide] Add hotkeys to show upload popup (#2073)
* [ide] Show a favicon in the browser IDE (#2070)
* [ide] Restyle bus nodes to avoid misclicks (#2069)
* [ide] Add hotkeys to create terminal(`t`), constant(`c`), tweak and watch(`w`) nodes while linking (#2068)
* [ide] Focus on label editor in Inspector when double-clicking terminal or bus nodes (#2063)
* [ide] Move to Electron v10, node.js v12 (#2084)

### Bug fixes

* [ide] Fix workspace errors when user deleted a workspace (#2082)
* [ide] Paste "expanded" nodes (such as constants, watches, tweaks) to the right position (#2071)
* [ide] Fix hotkeys not working in some cases (#2067)
* [ide] Show create new patch popup even when the project browser is hidden (#2066)
* [nodes] Fix a typo in welcome-to-xod project (#2064)

<a name="0.36.1"></a>

## 0.36.1 (2020-10-30)

### Bug fixes

* [nodes] Fix `xod/debug/is-receiving` implementation (#2061)
* [nodes] Revert `xod/core/animation-unit` to use `setTimeout` API (#2061)

<a name="0.36.0"></a>

## 0.36.0 (2020-10-29)

### Features and enhancements

* [core] Introduce records — an easy way to pack multiple data fields into a custom type and to serialize it to JSON. Check out [the guide article](https://xod.io/docs/guide/records/) for details (#2051)
* [c++] Introduce [`setImmediate`](https://xod.io/docs/reference/node-cpp-api/#setImmediate) API, that greatly optimises cases that require sheduling evaluation right after the current transaction completes (#2056)
* [ide] Add information about constant types in C++ cheatsheet in the sidebar (#2045)

### Bug fixes

* [nodes] Make xod graphics bitmaps compile on non-avr boards (#2053)
* [nodes] Fix `xod-dev/esp8266-mcu/lan-ip` implementation (#2050)

### New nodes

* Add `to-json` abstract node and specialisations for primitive data types to [xod/json](https://xod.io/libs/xod/json) library (#2051)

<a name="0.35.3"></a>

## 0.35.3 (2020-09-30)

### Bug fixes

* [c++] Fix dirtiness propagation of outputs of tweak nodes (#2048)

<a name="0.35.2"></a>

## 0.35.2 (2020-09-18)

### Features and enhancements

* [nodes] Ensure compatibility `xod-dev/servo` and `xod/gpio` with ESP32 boards (#2038)
* [ide] Make possible to send char literals to `tweak-byte`s in live session (#2041)
* [c++] Optimise memory consumption of `tweak-string` nodes (#2042)

### Bug fixes

* [ide] Fix loading of a newly added package index file in extra.txt (#2038)
* [ide] Fix hotkeys in Inspector Pin widgets (#2040)
* [nodes] Fix `tweak-string` nodes (#2042)

<a name="0.35.1"></a>

## 0.35.1 (2020-09-15)

### Bug fixes

* [core] Make sure implementation keywords are always detected correctly (#2035)
* [c++] Fix TetheringInternet namespace in generated code (#2036)
* [nodes] Fix typo in `xod/debug/get-local-ip` implementation (#2036)
* [nodes] Fix typo in `xod/common-hardware/ds18b20-thermometer` implementation (#2034)
* [nodes] Fix typo in `xod/graphics/humidity-16x16-rgba` implementation (#2032)

<a name="0.35.0"></a>

## 0.35.0 (2020-09-08)

### Features and enhancements

* [core] Introduce constant types and new C++ implementation syntax. (#2007)
* [core] Make color-type outputs bindable (#2011)
* [ide] Add ability to clone a patch from library to a local project (#1999, #2003)
* [nodes] Make `pulse-on-*` nodes behave consistently and not trigger on start (#2002)
* [core] Upgrade bundled arduino-cli to v0.12.0. Thanks, @quinkennedy! (#1986, #2012, #2029)
* [core] Use upstream ESP8266 package index instead of being locked to v2.4.3 (#2026)
* [core] Improved resolution of abstract nodes to allow specializations without the type suffix for types defined in the same library (#2004)
* [nodes] Unify interfaces of nodes in `xod-dev/esp8266`, `xod-dev/esp8266-mcu`, `xod-dev/w5500` and `xod/debug` to use abstract nodes from `xod/net` (#2004)

### Bug fixes

* [ide] Fix inconsistent behaviour of "Follow Log" feature in Debugger pane (#2027)
* [nodes] Fix compiling of `xod-dev/ws2812` on ESP8266 (#1994)
* [core] Fix an uncaught exception when transpiling a patch with a “dangling” defer node (#1995)
* [c++] Fix forward declaration of HardwareSerial for megaAVR architecture (#2006)
* [core] Fix creating cast nodes for custom types (#2008, #2023, #2025)

graph.js

var settings = {
    node: {
        width: 150,

        title: {
            height: 30,
        },

        endpoint: {
            radius: 3,
            step: 15,
            vmargin: 15,
            hpadding: 10,
        },
    }
}

var nodeTypes = {
    button: {
        title: 'Button',

        inputs: [
            'enabled',
        ],

        outputs: [
            'press',
            'release',
        ],
    }
}

var nodes = [{
    type: 'button',

    x: 10,
    y: 10,
}]


function nodeHeight(node) {
    var ntype = nodeTypes[node.type];
    var pcount = Math.max(ntype.inputs.length, ntype.outputs.length);
    return settings.node.title.height +
        2 * settings.node.endpoint.vmargin +
        (pcount - 1) * settings.node.endpoint.step;
}

function buildEndpoint(g, x, labelDx, labelAnchor) {
    var yBase = settings.node.title.height +
                settings.node.endpoint.vmargin;

    calcY = function(d, i) {
        return yBase + i * settings.node.endpoint.step;
    }

    g.append('circle')
        .attr('cx', x)
        .attr('cy', calcY)
        .attr('r', settings.node.endpoint.radius)
        .attr('fill', 'black');

    g.append('text')
        .text(function(d) { return d; })
        .attr('text-anchor', labelAnchor)
        .attr('dominant-baseline', 'central')
        .attr('x', x + labelDx)
        .attr('y', calcY);
}

function buildInput(g) {
    buildEndpoint(g, 0, settings.node.endpoint.hpadding, 'start');
}

function buildOutput(g) {
    buildEndpoint(g, settings.node.width,
                  -settings.node.endpoint.hpadding, 'end');
}

function buildNode(g) {
    g.attr('transform', function(d) {
        return 'translate(' +
            (Math.floor(d.x) + 0.5) + ', ' +
            (Math.floor(d.y) + 0.5) + ')';
    });

    g.append('rect')
        .attr('class', 'outline')
        .attr('width', settings.node.width)
        .attr('height', function(d) { return nodeHeight(d); })

    g.append('text')
        .text(function(d) { return nodeTypes[d.type].title; })
        .attr('text-anchor', 'middle')
        .attr('dominant-baseline', 'central')
        .attr('x', settings.node.width / 2)
        .attr('y', settings.node.title.height / 2)
        .attr('class', 'title')

    g.append('path')
        .attr('class', 'outline')
        .attr('d', 'm 0 ' + settings.node.title.height + ' h ' + settings.node.width)

    var input = g.selectAll('g.input')
        .data(function(d) { return nodeTypes[d.type].inputs; })
        .enter().append('g').attr('class', 'endpoint input');

    buildInput(input);

    var output = g.selectAll('g.output')
        .data(function(d) { return nodeTypes[d.type].outputs; })
        .enter().append('g').attr('class', 'endpoint output');

    buildOutput(output);
}

$(function() {
    var body = d3.select("body");

    var svg = body.append('svg')
        .attr('id', 'canvas')
        .attr('height', 600)
        .attr('width', 600);

    var node = svg.selectAll("g.node")
        .data(nodes)
        .enter().append("g").attr('class', 'node');

    buildNode(node);
});
